File: apc16delft_motion/src/master_pose.cpp
#include "master_pose.hpp"

namespace apc16delft_msgs {
	bool operator< (apc16delft_msgs::MasterPoseDescriptor const & a, apc16delft_msgs::MasterPoseDescriptor const & b) {
		return apc16delft::masterPoseToString(a) < apc16delft::masterPoseToString(b);
	}
}

namespace apc16delft {

std::string masterPoseToString(int const & bin_index) {
		return "bin" + std::to_string(bin_index);
}

std::string masterPoseToString(apc16delft_msgs::MasterPoseDescriptor const & descriptor) {
	if (descriptor.type == apc16delft_msgs::MasterPoseDescriptor::TYPE_BIN || descriptor.type == apc16delft_msgs::MasterPoseDescriptor::TYPE_CAM) {
		return descriptor.type + std::to_string(descriptor.bin_index);
	} else if (descriptor.type == apc16delft_msgs::MasterPoseDescriptor::TYPE_TOTE && descriptor.bin_index > 0) {
		int index = 0;
		std::string orientation = "";

		index = descriptor.bin_index - 1;
		orientation = (descriptor.group_name == descriptor.GROUP_TOOL1 ? "v" : "h");
		int x = index / 3;
		int y = index % 3;
		return descriptor.type + std::to_string(x) + std::to_string(y) + orientation;
	} else {
		return descriptor.type;
	}
}

}

File: apc16delft_coordinator/scripts/lib/master_pose.py
from apc16delft_msgs import msg

def motionGroupFromGraspStrategy(strategy):
	if strategy == msg.GraspCandidate.STRATEGY_VACUUM_V: return msg.MasterPoseDescriptor.GROUP_TOOL1
	if strategy == msg.GraspCandidate.STRATEGY_VACUUM_H: return msg.MasterPoseDescriptor.GROUP_TOOL0
	assert False, 'Invalid grasp strategy'

def home(motion_group = msg.MasterPoseDescriptor.GROUP_TOOL0):
	return msg.MasterPoseDescriptor(
		type         = msg.MasterPoseDescriptor.TYPE_HOME,
		group_name = motion_group,
		bin_index    = 0
	)

def bin(bin_index, motion_group = msg.MasterPoseDescriptor.GROUP_TOOL0):
	return msg.MasterPoseDescriptor(
		type         = msg.MasterPoseDescriptor.TYPE_BIN,
		group_name   = motion_group,
		bin_index    = bin_index
	)

def cam(bin_index, motion_group = msg.MasterPoseDescriptor.GROUP_TOOL0):
	return msg.MasterPoseDescriptor(
		type         = msg.MasterPoseDescriptor.TYPE_CAM,
		group_name   = motion_group,
		bin_index    = bin_index
	)

def tote(cell = 0, motion_group = msg.MasterPoseDescriptor.GROUP_TOOL0):
	return msg.MasterPoseDescriptor(
		type         = msg.MasterPoseDescriptor.TYPE_TOTE,
		group_name   = motion_group,
		bin_index    = cell
	)
